### PR TITLE
Customizable reject reason code

### LINKED ISSUE #1156: [FR] Enhance Rejection reason codes
### Use Case

1. Add `srt_setrejectionreason(..)` to make it possible for a listener side to set a custom rejection reason code in the listener callback.
2. Use this explicitly provided rejection reason code as a rejection reason ID, sent back to the caller in the Handshake Type field of the handshake packet.
What to do with SRT internal rejection reason? E.g. the passphrase provided by the listening side does not match the one provided by the caller.
In this case, SRT should provide its internal rejection reason (`SRT_REJ_BADSECRET`) and ignore a user-provided rejection reason code if any.

3. The caller is notified that an error occurred. It gets the rejection reason code provided by the listener using `srt_getrejectreason(..)` API function.

### Rejection Reason Codes

The Handshake Type field is reused, which means existing Handshake Type values have to be avoided [[RFC](https://haivision.github.io/srt-rfc/draft-sharabayko-mops-srt.html#handshake-type)].

Three spaces for rejection reason codes are defined:

1. SRT Internal Codes: rejection reason codes in case SRT rejects the connection by itself (e.g. in case of a mismatch in Congestion Control algorithms).

2. Predefined Codes. These rejection codes can be set by an upstream application through the listener callback, but they are defined (registered) by SRT.

3. User-defined Codes: rejection reason codes provided through the listener callback and not defined (registered) by SRT.

**SRT internal** rejection reason codes, use the following ranges:

- SRT API: [0; 1000) as declared in `enum SRT_REJECT_REASON` of the API. E.g. `REJ_UNKNOWN(0)`
- Protocol: [1000; 2000) as transmitted in the Handshake Type field of the CONCLUSION handshake. The actual value is `SRT_REJECT_REASON + 1000` (`1000 + REJ_UNKNOWN(0)`)

**SRT Predefined** Rejection Reason

- SRT API: [1000; 2000). Partially declared in the API. Mainly HTTP status codes are reused.
- Protocol: [2000; 3000) as transmitted in the Handshake Type field of the CONCLUSION handshake.

**SRT User Defined** Rejection Reason

- SRT API: [2000; ...). Not declared in the API.
- Protocol: [3000; ..) as transmitted in the Handshake Type field of the CONCLUSION handshake.

E.g. HTTP status code 406 Not Acceptable is 2000 + 406 = 2406 as transmitted by the protocol, and 1000 + 406 = 1046 as retrieved via `srt_getrejectionreason(..)`.

SRT Guidelines: reuse existing HTTP Status codes when possible.

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #1156

### CODE REVIEW

File: srtcore/common.cpp [hunk 1]
```diff
@@ -547,9 +547,16 @@ extern const char* const srt_rejectreason_msg [] = {
     "Group settings collision"
 };
 
-const char* srt_rejectreason_str(SRT_REJECT_REASON rid)
+const char* srt_rejectreason_str(int id)
 {
-    int id = rid;
+    if (id > SRT_REJC_SERVER)
```
Comment (maxsharabayko) [inline srtcore/common.cpp:552]: Maybe `>=`?

File: srtcore/common.cpp [hunk 2]
```diff
@@ -547,9 +547,16 @@ extern const char* const srt_rejectreason_msg [] = {
...
+    {
+        if (id > SRT_REJC_USER)
```
Comment (maxsharabayko) [inline srtcore/common.cpp:554]: `>=`

File: srtcore/handshake.cpp [hunk 3]
```diff
@@ -141,9 +141,22 @@ std::string RequestTypeStr(UDTRequestType rq)
...
-        int id = rej;
-        return std::string("ERROR:") + srt_rejectreason_name[id];
+        std::ostringstream rt;
+        rt << "ERROR:";
+        int id = RejectReasonForURQ(rq);
+        if (id < SRT_REJ_E_SIZE)
+            rt << srt_rejectreason_name[id];
+        else if (id < SRT_REJC_USER)
+        {
+            if (id < SRT_REJC_SERVER)
```
Comment (maxsharabayko) [inline srtcore/handshake.cpp:151]: `<=` instead of `<`?

File: srtcore/handshake.h [hunk 4]
```diff
@@ -237,25 +237,36 @@ enum UDTRequestType
...
 {
-    if (req < URQ_FAILURE_TYPES || req - URQ_FAILURE_TYPES >= SRT_REJ__SIZE)
+    if (req < URQ_FAILURE_TYPES)
         return SRT_REJ_UNKNOWN;
-    return SRT_REJECT_REASON(req - URQ_FAILURE_TYPES);
+
+    if (req < URQ_SERVER_FAILURE_TYPES && req - URQ_FAILURE_TYPES >= SRT_REJ_E_SIZE)
+        return SRT_REJ_UNKNOWN;
+
+    return req - URQ_FAILURE_TYPES;
```
Comment (maxsharabayko) [inline srtcore/handshake.h:269]: Can be simplified as:
```suggestion
    const int reason = req - URQ_FAILURE_TYPES;
    if (reason >= SRT_REJ_E_SIZE)
        return SRT_REJ_UNKNOWN;

    return reason;
```

File: srtcore/srt.h [hunk 5]
```diff
@@ -573,9 +573,17 @@ enum SRT_REJECT_REASON
     SRT_REJ_FILTER,      // incompatible packet filter
     SRT_REJ_GROUP,       // incompatible group
 
-    SRT_REJ__SIZE,
+    SRT_REJ_E_SIZE,
```
Comment (maxsharabayko) [inline srtcore/srt.h:576]: API change. Someone could have already used `SRT_REJ__SIZE`.

File: srtcore/srt.h [hunk 6]
```diff
@@ -573,9 +573,17 @@ enum SRT_REJECT_REASON
...
 };
 
+// Reject category codes:
+
+#define SRT_REJC_VALUE(code) (1000 * (code/1000))
+#define SRT_REJC_SYSTEM 0     // Codes from above SRT_REJECT_REASON enum
+#define SRT_REJC_SERVER 1000  // Standard server error codes
+#define SRT_REJC_USER 2000    // User defined error codes
```
Comment (maxsharabayko) [inline srtcore/srt.h:584]: I don't think we need to distinguish SERVER and USER. All these rejection reason codes are user-defined.

File: srtcore/handshake.h (hunk 4 above)
Comment (ethouris) [reply srtcore/handshake.h:269]: This changes the behavior for the range between `SRT_REJ_E_SIZE` and `SRT_REJC_SERVER`. But the use of a separate variable makes things clearer, I'll fix it that way.

File: srtcore/srt.h (hunk 5 above)
Comment (ethouris) [reply srtcore/srt.h:576]: I'll make a `#define`. This label must be deprecated - symbols with double `_` inside are reserved for the compiler - I've learned that quite late.

File: srtcore/srt.h (hunk 6 above)
Comment (ethouris) [reply srtcore/srt.h:584]: That's a design-level discussion and we're past that point already.

File: srtcore/handshake.cpp (hunk 3 above)
Comment (ethouris) [reply srtcore/handshake.cpp:151]: I thought your previous comment defined that, e.g., `SRT_REJC_SERVER` exact value is included in the range for server codes...

File: srtcore/handshake.h [hunk 7]
```diff
@@ -263,10 +263,11 @@ inline int RejectReasonForURQ(UDTRequestType req)
     if (req < URQ_FAILURE_TYPES)
         return SRT_REJ_UNKNOWN;
 
-    if (req < URQ_SERVER_FAILURE_TYPES && req - URQ_FAILURE_TYPES >= SRT_REJ_E_SIZE)
+    int reason = req - URQ_FAILURE_TYPES;
+    if (reason < SRT_REJC_SERVER && reason >= SRT_REJ_E_SIZE)
```
Comment (maxsharabayko) [inline srtcore/handshake.h:267]: I may assume `reason >= SRT_REJ_E_SIZE` already covers `reason < SRT_REJC_SERVER`

File: srtcore/srt.h (hunk 6 above)
Comment (maxsharabayko) [reply srtcore/srt.h:587]: We are exactly at that point now, we are trying to provide a useful API, not just merge something and put a checkbox.

Comment (ethouris) [reply srtcore/srt.h:587]: So, the merit was the following: the rejection code is something that one application sends and the other - usually completely unrelated application - interprets. Therefore there are reserved two spaces, one for HTTP codes, one for user codes. When an application uses HTTP codes, it should align to the requirements of the HTTP codes so that the other - unrelated - application understands it.

If then an application is using the error code from the "server" range, it must use HTTP codes, or risk being falsely understood as not conforming to the standard. If the application uses the "user" range however, it must make sure - by interchanging some other data using possibly another connection or another medium - that the application "speaks the same language" before specifying any code like this.

In short:

"server" codes: all applications _should_ speak the same language (we can't enforce that of course, but we can at least define a standard).

"user" codes: applications on both sides must be somehow certain, or make sure somehow, that they both speak the same language, and avoid using these codes otherwise (or reject a connection in advance using a HTTP code).

File: srtcore/handshake.h (hunk 7 above)
Comment (ethouris) [reply srtcore/handshake.h:267]: Excuse me?

It simply maps to `reason >= 16` and it should cover `reason < 1000` ? How about the values of `reason >= 1000` ?

Comment (ethouris) [reply srtcore/handshake.h:267]: In short: the rule here is that:
 - for rejection codes between 1 and `SRT_REJ_E_SIZE-1` the rejection value is returned
 - for values from `SRT_REJC_SERVER` on, the rejection value is returned (even if this results in something stupid, like an nonexistent HTTP code)
 - otherwise (0 and between `SRT_REJ_E_SIZE` and `SRT_REJCT_SERVER-1`) it should return UNKNOWN

Comment (maxsharabayko) [reply srtcore/handshake.h:267]: Right, I see what you are doing here now.
You are returning `SRT_REJ_UNKNOWN` for those values, that are right after `SRT_REJ_E_SIZE` (including) and up to first user-defined code `SRT_REJC_SERVER`.

File: srtcore/handshake.cpp (hunk 3 above)
Comment (maxsharabayko) [reply srtcore/handshake.cpp:151]: Right, all good. The function returns "unknown" if `id < SRT_REJC_SERVER`

File: srtcore/handshake.h (hunk 7 above)
Comment (ethouris) [reply srtcore/handshake.h:267]: Exactly. This is returned for UNKNOWN itself and for undefined codes in the SRT internal range only. In the server and user ranges it is returned as is.
